### DISCUSSION TITLE
Import Solution - "Upgrade"

### DISCUSSION THREAD

**Discussion by fowl2** [microsoft/powerplatform-build-tools#494] [Ideas]
Dataverse has this 'wonderful' choice between "Upgrade", "Stage for Upgrade" (aka "Holding"^), and "Update". 

The CLI currently has commands for the last two, but not the first.

So we have the choice of either:

| Method | Pros | Cons |
|--------|------|------|
| `pac solution import` | Works regardless of whether  solution is already installed. | Won't clean up deleted components. |
| `pac solution import --import-as-holding` <br> and then <br> `pac solution upgrade` | Cleans up deleted components. | Will error if the solution isn't already installed. |

It'd be awesome if we had a `pac solution import --upgrade` that handled both situations in one step.

I could also see adding a file path parameter to `pac solution upgrade` if for some reason you wanted an error/warning/skip if the solution wasn't already installed might be useful for some.

-------------

Refresher on the difference between "Upgrade", "Stage for Upgrade" ("Holding"^), "Update":

> **Upgrade** This is the default option and upgrades your solution to the latest version and rolls up all previous patches in one step. Any components associated to the previous solution version that are not in the newer solution version will be deleted. This option will ensure that your resulting configuration state is consistent with the importing solution including removal of components that are no longer part of the solution.
> 
> **Stage for Upgrade** This option upgrades your solution to the higher version, but defers the deletion of the previous version and any related patches until you apply a solution upgrade later. This option should only be selected if you want to have both the old and new solutions installed in the system concurrently so that you can do some data migration before you complete the solution upgrade. Applying the upgrade will delete the old solution and any components that are not included in the new solution.
> 
> **Update** This option replaces your solution with this version. Components that are not in the newer solution won't be deleted and will remain in the system. Be aware that the source and destination environment may differ if components were deleted in the source environment. This option has the best performance by typically finishing in less time than the upgrade methods.

-- https://learn.microsoft.com/en-us/power-apps/maker/data-platform/update-solutions

^ PS I guess some consistency on "stage" vs "holding" might be nice too :)

Thanks!

edit: this is implemented as `--stage-and-upgrade`! Would still be nice to have "install-or-stage-and-upgrade".

**Comment by YannickRe** [comment]
Seems like this one, #385 , and #501 all highlight the same (or at least a similar problem).

**Comment by lewginn** [comment]
I think the Upgrade import method would be a good start to help with Upgrade/Staged deployments. The "skip lower version" toggle is only partly working at the minute, especially when it comes to applying solution upgrades. It can skip over the Import as holding solution part fine, but skipping over an item that doesn't have an upgrade available?

Having the upgrade option saves a load of time when importing solutions manually, so surely it's on the list for automated deployments!

**Reply by YannickRe** [reply]
Yes, as the Holding + Apply is just double as slow as the UI version of "Upgrade" that does it all in one go. The team should look at it holistically:

- Add an equivalent to the UI "upgrade" action
- Make sure we can do an import/upgrade that handles all of it: import new, update, upgrade, skip-lower-version so that reruns of the pipeline with the same version solution don't fail, but gracefully continue.

We currently solve this with a PowerShell script that calls the environment API, checks for existing solutions and their versions, and determines the config for the subsequent steps. Works great now, but would be nicer if we have it from PAC CLI natively.

**Reply by lewginn** [reply]
Couldn't agree more. I think that's essentially what the Power DevOps toolset was doing to. Which I think I'll stick with until the upgrade feature is released

**Comment by jcoxdco** [comment]
@petrochuk, is this something that the team will be able to take up in the near future?

**Reply by petrochuk** [reply]
yes, we are making changes in solution import area

**Reply by MattB-msft** [reply]
We are also in flight on a pac package deploy update to accept a single solution, this will apply the Package deployer logic to update or upgrade the package.

**Reply by ciaomatteo** [reply]
Which version of solution upgrade is package deployer using?

**Comment by snizar007** [comment]
@fowl2, @lewginn, @YannickRe, @jcoxdco - We have release new version of Power Platform CLI (PAC CLI) ver: 1.28.x where we have introduced pac solution import with --stage-and-upgrade. Please try and provide feedback.

![image](https://github.com/microsoft/powerplatform-build-tools/assets/116691001/a29c12a8-8fad-4f56-961d-5e70172219e8)

**Reply by jcoxdco** [reply]
Thanks @snizar007  I love the new functionality. Works great in conjunction with the --skip-lower-version flag.

The issue I'm running into is that *each* task (Import and then Upgrade) had a two hour time limit enforced. Now the single task of Import Solution with the stage-and-upgrade flag still has a total of 2 hours. My largest solution currently takes about 2 hours 15 minutes to import and upgrade which just results in DevOps reporting an error due to the 2 hour hard limit. (I realize I also have some work to do to try to carve this solution up even more to try to get the overall import times down.)

There a reason that time limits are enforced in the PPBT tasks but are different than the experience when using make.powerapps.com?

**Reply by snizar007** [reply]
have you tried using --async with your command?

**Reply by jcoxdco** [reply]
The error is for sure PEBKAC. I remembered to change the max wait time but forgot to update the timeout under the control options. Sorry to waste your time. deployments over two hours (although not great) are supported in the tools. Thanks.

**Comment by YannickRe** [comment]
@snizar007 Thank you for this, and I tried it out.
Unfortunately, adding the extra switch just increased confusion for me:

- What's the behavior when I combine `--import-as-holding` and `--stage-and-upgrade`? Either having them both as `true`, both as `false`, and one `true` and the other `false`? This doesn't seem documented.
- How does this switch behave in combination with `--skip-lower-version`?
- How does this switch work when the solution is net new on the target environment? Can I do `--skip-lower-version true --stage-and-upgrade true` and make it import new solutions and upgrade existing solutions? What would be the one command I can use for both new installs as stage and upgrade for existing solutions, assuming I do not know the state of the current solution on that specific environment?
- This command isn't faster: it is now taking the same time as an "import as holding" + "apply solution upgrade" task, whereas running "upgrade" through the UI is only half that time. We are getting it as one, but not getting the speed advantages you have in the UI when you do it as one, while increasing confusion when combining with existing flags...

**Reply by YannickRe** [reply]
@snizar007 do you have any clarifications here? Thank you!

**Reply by tchinnin** [reply]
I'm facing same problem here on Power Platform Build Tools for AzureDevOps 2.0.49

Using `--skip-lower-version true` and/or `--stage-and-upgrade true` fail the import if solution does not exist on target environment.

At some point I'll have to perform same as you do @YannickRe  : Developing custom PowerShell/PAC function that checks if solution exists in target environment to properly set `--skip-lower-version` and/or `--stage-and-upgrade` solution import parameters.

Would be great if we had an out-of-the-box way to deal with new solutions without adding extra steps.

**Reply by nguyentranenvoyair** [reply]
As a workaround, you can try using "continue-on-error: true" in your step. If solution does not exist in target environment, the step returns an error and you can retry that same step, this time giving it different parameter values.

For more on how to retry a step, please look at this [discussion](https://github.com/orgs/community/discussions/27121)

**Comment by crsnt** [comment]
I've been experiencing frustrations with large project deployments as well. 
Import and Upgrade takes too long, sometimes the next solution import failed because the previous solution hasn't finished upgrading (even though DevOps said its finished).
 
Environments are not consistent either. Some environments seem to process quicker and some take 3 times slower even though they have same configuration, which is weird and I have ongoing ticket with MS.

Would really love major improvement on solution deployment on how to make it faster. Total deployment time can take 4-7 hours!

**Comment by ADMIN-SKE2388** [comment]
I just saw this video and noticed behavioural changes when I upgrade CoE. It does not show both solution before it uninstall the old one like in past. Now it seems that "upgrade" more like in line update. Does anyone notice the same thing?

https://learn.microsoft.com/en-us/shows/the-low-code-revolution/importing-managed-solutions-what-has-changed-to-the-upgrade-verses-update-process
